azure/Kqlmagic/adal_token_cache.py

```diff
@@ -117,14 +176,116 @@ def _query_cache(self, is_mrrt, user_id, client_id):
                 matches.append(v)
         return matches
 
+    def clear_db(self):
+
+        items = self.db.items()
+        
+        for token_tuple in items:
+
+            db_key = token_tuple[0] # 'kqlmagicstore/tokens/cachename' address
+
+            cachename = db_key[len(AdalTokenCache.key_prefix_db):]
+            data = token_tuple[1] #encrypted token
+            if not data:
+                return
+            try:
+                time_created, _unused = Fernet._get_unverified_token_data(data)  #get token timestamp 
+            except InvalidToken:
+                Display.showWarningMessage("Warning: found an illegal token, deleting the token")
+                del self.db[f"{AdalTokenCache.key_prefix_db}{cachename}"]
+            EXP_TIME = self.token_exp_time  
+            if(time_created +EXP_TIME < int(time.time()) ):
+                del self.db[f"{AdalTokenCache.key_prefix_db}{cachename}"]
+
     def _save(self):
         data = self.serialize()
-        self.db['kqlmagicstore/tokens'] = data
+        data_as_bytes = data.encode()
+
+        #init fernet
+        fernet = self.fernet
+
+        #encrypt 
+        data_encrypted = fernet.encrypt(data_as_bytes)
+        
+        self.db[f"{AdalTokenCache.key_prefix_db}{self.cachename}"] = data_encrypted
 
     def _restore(self):
+
+        if(AdalTokenCache.last_clear_time + TIME_BETWEEN_CLEANUPS < int(time.time()) ):
+            AdalTokenCache.last_clear_time = int(time.time())
+            self.clear_db()
+
+        data_encrypted = self.db.get(f"{AdalTokenCache.key_prefix_db}{self.cachename}")
+        if not data_encrypted:
+            return
+        fernet = self.fernet
         try:
-            data = self.db['kqlmagicstore/tokens']
-        except KeyError:
-            print("no stored tokens")
+            data_decrypted_as_bytes = fernet.decrypt(data_encrypted)   
+        except InvalidToken: #Either token has bad form or it cannot be decrypted
+            try: 
+                Fernet._get_unverified_token_data(data_encrypted)
+            except: #the token has bad form
+                Display.showWarningMessage("Warning: found an illegal token, deleting the token")
+                del self.db[f"{AdalTokenCache.key_prefix_db}{self.cachename}"]
+                return
+            Display.showWarningMessage("Invalid token, could not activate Single Sign On") #the token cannot be decrypted
+            return
+        data = data_decrypted_as_bytes.decode()
+        self.deserialize(data)
+
+    @staticmethod
+    def get_params_SSO(**options): #pylint: disable=no-method-argument
+
+        SSO_id_enc_key = os.getenv("{0}_SSO_ENCRYPTION_KEYS".format(Constants.MAGIC_CLASS_NAME.upper()))
+        key_vals_SSO = Parser.parse_and_get_kv_string(SSO_id_enc_key, {}) if SSO_id_enc_key else {}
```

Comment:
I am missing a check if SSO_id_enc_key is None:
    Display.showWarningMessage(f"Warning: SSO is not activated because environment variable {Constants.MAGIC_CLASS_NAME.upper()}_SSO_ENCRYPTION_KEYS is not set"